File: Data Structures & Algorithms/validate-parentheses/submission-4.py
'''
input: string of chars
want: boolean representing if all pairs have a corresponding opening and closing pair

edge cases: only brackets right? no other chars, empty input

time, space - 

exp:

s = |    ([{}])    |
pairs = {')' : '('}

for each value we add closing brackets to the stack and if we see an opening we pop it and return of the len(stack) == 0

if its an opening bracket and hte last element on hte stack does match (corresponds pairs[s[i]]) then we pop if not we append it













'''
class Solution:
    def isValid(self, s_str: str) -> bool:
        stack = []
        pairs = {')': '(', ']':'[', '}':'{'}

        if len(s_str) % 2 != 0: return False # odd len will never be true


        for char in s_str:
            if char in pairs and stack and stack[-1] == pairs[char]: stack.pop()
            else:
                stack.append(char)
        return True if len(stack) == 0 else False

File: Data Structures & Algorithms/duplicate-integer/submission-4.cpp
/*
in: array of ints 
out: boolean, if dups present return true otherwise return false
edge case: empty array in input

time, space - 

solultion
we add all the elements to a set then if the length of the nums array does not match the length of the set then we have duplicates present so we return true
otherwise false


*/



class Solution {
public:
    bool hasDuplicate(vector<int>& nums) {
       if (nums.empty()) return false;

       unordered_set<int> seen (nums.begin(), nums.end());
       return nums.size() != seen.size();
    }
};